[In progress] Add a Carousel block

philcable opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Carousel block (Images, Posts, Cards, Products) plus Navigation and Pagination blocks with drag/swipe, keyboard navigation, accessible controls, animations, customizable styles, and WooCommerce product support; automatic update integration included.
* **Documentation**
  * Added plugin readme and Extending guide for integrating custom track blocks and item-count behavior.
* **Style**
  * Added editor and front-end styles for carousel, navigation, and pagination.
* **Chores**
  * Added editorconfig, gitignore, package metadata, build scripts, and lint config.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

philcable committed:
Add initial styles for gallery carousels

These may seem heavy-handed, but they're the lightest I was able to come up with while still overriding some competing core styles.

philcable committed:
Add initial handling for uncropped galleries

philcable committed:
Add initial carousel content selection interface

philcable committed:
Add initial block controls

This also adds some very rough styles and a view script that doesn't do anything yet.

philcable committed:
Add initial front-end functionality

philcable committed:
Fix default item gap

philcable committed:
Merge pull request #51 from a8cteam51/feature/17-gallery

Add initial gallery carousel handling

nate-allen commented on carousel/src/common.js:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	return {
+		'aria-label': 'block title', // TODO: Add block title.
```
Add block title

nate-allen commented on carousel/carousel.php:
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Plugin Name:       Carousel
+ * Description:       Example block scaffolded with Create Block tool.
```
Please update the description

philcable replied:
Thanks for catching these, @nate-allen! I'll get both fixed up soon.

nate-allen reviewed (changes requested):
Just a couple minor things that need to be updated. Thank you

philcable committed:
Fix plugin description

philcable committed:
Add initial arrow button position handling

The "Left" position needs a bit more work. I don't want to get too hung up on that right now, though.

philcable committed:
Add PHP rendering

It's a bummer to have to replicate all this, but it affords the ability to change things without invalidating saved blocks.

philcable committed:
Merge pull request #53 from a8cteam51/feature/17-block-settings

Add initial arrow button position handling

philcable committed:
Add support for arbitrary cards

philcable committed:
Update itemCount logic to accommodate arbitrary cards

philcable committed:
Fix carousel block selectability

Adding a `ref` to the block itself interferes with its ability to be selected, so this shuffles things around a bit and adds an empty div for the sole purpose of holding the ref.

philcable committed:
Improve gap style handling

This applies `gap` only in cases where its needed so extra spacing isn't hanging out otherwise.

philcable committed:
Merge pull request #54 from a8cteam51/feature/17-cards

Carousel Block: Add support for arbitrary cards

philcable committed:
Clean up view script

philcable committed:
Merge pull request #56 from a8cteam51/feature/17-render

Carousel Block: Add PHP rendering

jeremyfelt committed:
Ensure all files are included on `wp-scripts plugin-zip`

Without this, the `classes/` directory is not included with the
build.

See https://developer.wordpress.org/block-editor/reference-guides/packages/packages-scripts/#plugin-zip

jeremyfelt committed:
Don't check wp.org for updates to this plugin slug

philcable committed:
Merge pull request #57 from jeremyfelt/fix/add-auto-update-class

Carousel: Prevent wp.org updates; package all files with `wp-scripts plugin-zip`

philcable committed:
Add initial drag/swipe handling

philcable committed:
Merge pull request #58 from a8cteam51/feature/17-swipe

Carousel Block: Add initial drag/swipe handling

philcable committed:
Add support for a title for screen reader users

philcable committed:
Merge pull request #59 from a8cteam51/fix/17-title

Carousel Block: improve title support for screen reader users

philcable committed:
Remove escaping from carousel content

This is redundant, and can break certain embeds, etc.

philcable committed:
Merge pull request #61 from a8cteam51/fix/17-content-escaping

Carousel Block: remove escaping from carousel content

philcable committed:
Add an animation speed control

philcable committed:
Merge pull request #62 from a8cteam51/feature/17-speed-option

Carousel Block: add an animation speed control

philcable committed:
Improve drag navigation handling

This adds logic to ensure we only navigate slides on generally horizontal swipes/drags,
and consolidates handling to make things more DRY.

philcable committed:
Merge pull request #63 from a8cteam51/fix/swipe-tolerance

Carousel block: improve drag navigation handling

philcable committed:
Add a control for selecting slideshow end behavior

philcable committed:
Add initial infinite scroll handling

This is rough and needs to be revisited, but may work well enough for now.

philcable committed:
Disable pagination buttons when infinite scroll is selected

We can revisit this later, too.

philcable committed:
Ensure transitionend listener is cleaned up after animating slides

philcable committed:
Merge pull request #64 from a8cteam51/feature/infinite-loop

Carousel block: add initial infinite loop handling

philcable committed:
Ensure that grid area style is applied to the correct selectors

philcable committed:
Disable "Go back to first slide" option until support for it is added

philcable committed:
Fix track selector for post carousels

philcable committed:
Improve column gap handling

philcable committed:
Merge pull request #65 from a8cteam51/fix/17-posts-track

Carousel block: fix posts track

philcable committed:
Fix undefined variable issue

philcable committed:
Merge pull request #66 from a8cteam51/fix/17-get-item-gap

Carousel block: fix undefined variable

philcable committed:
Ensure that children of non-visible slides can't be focused

philcable committed:
Ensure that focus is reset if lost during the slide animation

philcable committed:
Merge pull request #67 from a8cteam51/fix/17-slide-focus

Carousel block: address slide focus and focusability handling

philcable committed:
Refactor carousel nav buttons as a block

philcable committed:
Refactor carousel pagination buttons as a block

I would prefer to use the `SpacingSizesControl` component for button size management inside `<InspectorControls group="dimensions">` to better match default controls, but both it and the `ToolsPanelItem` wrapper are still experimental. Which would be fine if they worked - unfortunately, I was unable to get them to.

philcable committed:
Adjust styles to improve editor experience

philcable committed:
Add layout support to the carousel block

philcable committed:
Fix track container name

philcable committed:
Improve navigation and pagination block controls

philcable committed:
Ensure navigation works when prev/next buttons aren't in place

philcable committed:
Adjust alignment support

philcable committed:
Remove convulted gallery handling and introduce a track height control

The track height needs more careful consideration - should it apply just to gallery carousels?

philcable committed:
Update arbitrary card handling

philcable committed:
Update navigation and button block color controls

philcable committed:
Accommodate nesting

philcable committed:
Improve pagination button handling

philcable committed:
Display track height control only for gallery carousels

philcable committed:
Adjust selector for query and product collection display in editor

philcable committed:
Allow custom icons for the navigation buttons

philcable committed:
Merge pull request #68 from a8cteam51/feature/17-buttons

Carousel block: refactor controls as separate blocks

philcable committed:
Fix check for product collection block

philcable committed:
Fix selector for Product Collection block

philcable committed:
Merge pull request #72 from a8cteam51/feature/17-woocommerce-quick-fixes

Carousel block: refine WC product collection block handling

philcable committed:
Add back to beginning/end option

philcable committed:
Divest from experimental components for track height control

philcable committed:
Update NPM packages

philcable committed:
Update readme.txt

philcable committed:
Add styles for left and right overflow options

philcable committed:
Validate JSON and check required fields in self_update() function

philcable committed:
Update block description and readme file

philcable committed:
Incorporate coderabbit suggestions

Fix variable scoping in switch statement.
Fix typo in JSDoc comment.
Add null check for paginationButtons to prevent potential errors.
Remove duplicate event listeners on pagination buttons.
Handle non-numeric gap values to prevent NaN.
Replace findLast with a more compatible solution (findLast is not supported in Safari < 15.4 and some other browsers).

philcable committed:
Update disabled pagination button styles

The `not-allowed` cursor is kind of corny - this forgoes it in favor of being more consistent with the disabled nav button styles.

philcable committed:
Add handling for pagination button states

I think I'd avoided this because it gets a little weird when multiple slides are considered visible at once.

philcable committed:
Update Author and Update URI

philcable committed:
Refactor to better handle infinite mode

philcable committed:
Refactor view script

Implements a linked list, reorganizes, and renames functions in an attempt to improve scannability, readability, and maintainability.

philcable committed:
Fix clip path on overflow right rule

philcable committed:
Adjust flex rules for individual slides

philcable committed:
Merge pull request #84 from a8cteam51/feature/17-carousel-refac

Carousel block: refactor view script

philcable committed:
Implement a common selector for the carousel track

This involves making a decision for the Product Collection block.

philcable committed:
Merge pull request #102 from a8cteam51/fix/17-carousel-track-selector

Carousel block: Implement a common selector for the carousel track

philcable committed:
Fix Sticky handling in item count logic

philcable committed:
Ensure accuracy of the  `--item-count` CSS variable on the front-end

philcable committed:
Add a meta name to all track blocks

philcable committed:
Merge pull request #104 from a8cteam51/fix/17-item-count

Carousel block: fix item count handling

philcable committed:
Fix bugs and remove dead code

philcable committed:
Improve front-end i18n and accessibility

Ans some general cleanup

philcable committed:
Update Carousel Nav block icon

Still not great.

philcable committed:
Update NPM packages, add an eslint config

Evidently, the root ESLint config fails to load when run from subdirectories due to the `const { version } = require( './package' );` line.
This resolves to the working directory, which is correct for this project, but it causes ESLint to silenty fall back to the @wordpress/scripts default config.

philcable committed:
Support third-party blocks as carousel tracks

philcable committed:
Fix WooCommerce Product Collection variation handling

And add a note about how to use this same pattern to add other blocks to the Carousel's placeholder UI.

philcable committed:
Improve item count in editor for third-party carousel tracks

Add filterable item count resolvers and developer documentation for extending the carousel with custom block types.

philcable committed:
Merge pull request #124 from a8cteam51/feature/17-carousel-extensibility

Carousel block: improve extensibility

fmfernandes commented:
Hi, @philcable 👋🏻 

Could we rename the plugin from `carousel` to maybe something else? 🤔 

It's conflicting with a plugin on WordPress.org and giving a false positive, see #115.

philcable commented:
Hi @fmfernandes!

Absolutely! As this will likely become an issue with other blocks with general/generic names, we should probably come up with a standard prefix? Maybe `spbm-`, `t51-`, `sp-`, or `a8csp-`? I'll defer to the preference of whomever is overseeing the monorepo :).

nate-allen commented:
Let's go with `a8csp-` as the prefix, since [the documentation says](https://github.com/a8cteam51/special-projects-blocks-monorepo/wiki/Creating-a-New-Block#development-notes) we should be using that for name spacing anyway

philcable committed:
Fix carousel block slug and namespace

philcable committed:
Fix Query block variation

philcable committed:
Update files included in plugin zip

philcable committed:
Update changelog

philcable committed:
Update project version

philcable committed:
Merge pull request #146 from a8cteam51/fix/17-carousel-slug-and-ns

Carousel block: fix slug and namespace

fmfernandes commented:
Hey, @philcable, is this block ready for a review again?

philcable commented:
> Hey, @philcable, is this block ready for a review again?

@fmfernandes It is indeed!
